docs/hugo/content/design/reconcile-interface.md

```diff
@@ -0,0 +1,227 @@
+    }
+```
+
+In create/update:
+```go
+    reconcilePolicy := reconcilers.GetReconcilePolicy(user, log)
+    if !reconcilePolicy.AllowsModify() {
+        return r.handleSkipReconcile(ctx, log, user)
+    }
+```
+
+Most of the implementation of `handleSkipReconcile` is also duplicated. One gotcha is that, in ARM resources case we want to populate status as part of that call, which requires a GET call to ARM, but in other cases we may not want to do anything.
+
+### Claiming a resource
+
+```go
+    if r.NeedToClaimResource(user) {
+        // TODO: This assumes we're requeing and doing the "stages" thing, which I am not sure if it makes sense to do for everybody anymore
+        _, err := r.ClaimResource(ctx, log, user)
+        if err != nil {
+            return ctrl.Result{}, err
+        }
+    }
+```
+
+Some claiming stuff (adding the finalizer) is universal. Other claiming stuff isn't, so we'd need a hook here.
+
+### Checking if a resource has been claimed before deletion
+
+```go
+    // If we have no resourceID to begin with, or no finalizer, the Azure resource was never created
+    hasFinalizer := controllerutil.ContainsFinalizer(r.Obj, reconcilers.GenericControllerFinalizer)
+    resourceID := genruntime.GetResourceIDOrDefault(r.Obj)
+    if resourceID == "" || !hasFinalizer {
+        return ctrl.Result{}, r.RemoveResourceFinalizer(ctx, r.Log, r.Obj)
+    }
+
+```
+
+Note that this check may need to be expanded by individual reconcilers. For example the one above checks both resourceID (ARM specific) and the standard finalizer (generic)
+
+### Some logging/eventing
+
+```go
+    msg := "Starting delete of resource"
+    r.Log.V(Status).Info(msg)
+    r.Recorder.Event(r.Obj, v1.EventTypeNormal, string(DeleteActionBeginDelete), msg)
+```
+
+## Proposal
+
+We create a more detailed interface and do as much work as we can in the `GenericController`. Each of these proposed methods would be idempotent:
+
+```go
+type Reconciler interface {
+    CreateOrUpdate(
+        ctx context.Context,
+        log logr.Logger,
+        eventRecorder record.EventRecorder,
+        obj MetaObject) (ctrl.Result, error)
+
+    Delete(
+        ctx context.Context,
+        log logr.Logger,
+        eventRecorder record.EventRecorder,
+        obj MetaObject) (ctrl.Result, error)
+
+    Claim(
+        ctx context.Context,
+        log logr.Logger,
+        eventRecorder record.EventRecorder,
+        obj MetaObject) error // No ctrl.Result here. An error means we can't proceed, no error means proceed directly to CreateOrUpdate.
+
+    // This is only called by the GenericController in cases where CreateOrUpdate couldn't be called due to reconcile policy
+    UpdateStatus(
+        ctx context.Context,
+        log logr.Logger,
+        eventRecorder record.EventRecorder,
+        obj MetaObject) error // No ctrl.Result here. An error means we couldn't get the status. Generic Controller will 
```

Comment:
I've moved them, but the actual code will also have more detailed comments. The intent with this snippet here was just to convey the idea and rough semantics so that people could get a sense if they linked the approach or not. Since there seems to be relative agreement that the approach is worth pursing I'll turn it into a PR which will have a clean and robust set of comments that also document things like what the individual reconcilers should rely on the generic controller for.